### DISCUSSION TITLE
dreaming ideas for teaching / IPython / Thonny

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#355] [Ideas]
I was playing with IPython's "REPL", which is so nice, and thinking, how nice would it be if we could have IPython at Thonny's lower window slot? Would it be possible?

My harder to achieve dream, I'm afraid it might not be possible, would be to have non-blocking imported mode sketches running on Thonny and being able to interact with them in the console/REPL.

**Comment by hx2A** [comment]
> I was playing with IPython's "REPL", which is so nice, and thinking, how nice would it be if we could have IPython at Thonny's lower window slot? Would it be possible?
> 
> My harder to achieve dream, I'm afraid it might not be possible, would be to have non-blocking imported mode sketches running on Thonny and being able to interact with them in the console/REPL.

I don't know how to do that in Thonny, but you can use IPython with py5 imported mode. The command is `jupyter console`, which is basically the same as (but not identical to) what you get with the `ipython` command. Then add the `--kernel=py5` param to specify what kernel you want:

```txt
% jupyter console --kernel=py5                          
Jupyter console 6.6.3

Python 3.8.17 | packaged by conda-forge | (default, Jun 16 2023, 07:11:34) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.2 -- An enhanced Interactive Python. Type '?' for help.

py5 0.9.2.dev0 | py5 kernel 0.1.3a0 | A Python Jupyter kernel plus py5 in imported mode
In [1]
```

**Reply by villares** [reply]
this looks very promissing!
I might have to rebuild my conda env for py5...
![image](https://github.com/py5coding/py5generator/assets/3694604/b55292b4-6d86-428f-a6ed-1a772d395877)

**Reply by hx2A** [reply]
I don't know why that didn't run, but I would try rebuilding the conda environment and maybe install java with installjdk.

**Reply by villares** [reply]
UPDATE: ignore this...  I forgot about `python -m py5jupyter.kernels.py5.install --sys-prefix` ... the error message could be more helpful I guess... `No module named 'py5jupyter'` makes one think one needs to install `py5jupyter` that is already installed...

Quick question, as I created a new conda env using `conda env create -n py5coding -f http://py5coding.org/files/install/py5_environment.yml`

Per this:
![image](https://github.com/py5coding/py5generator/assets/3694604/1f34e648-9b4a-4d91-9ec4-c56d6a425a09)

Should it install `py5jupyter`? It didn't work for me :(
```shell
[villares@L340 ~]$ conda activate py5coding
(py5coding) [villares@L340 ~]$ jupyter console --kernel=py5
/usr/bin/python: Error while finding module specification for 'py5jupyter.kernels.py5' (ModuleNotFoundError: No module named 'py5jupyter')
``` 
After `pip install py5jupyter` told me all requirements already satisfied, but still no luck.

**Reply by villares** [reply]
OHHH! I forgot about `python -m py5jupyter.kernels.py5.install --sys-prefix`

**Comment by villares** [comment]
This is about dreaming right? It`s about thinking of the most magic & wonderful tools possible without prejudice?

So, after installing `py5jupyter` could we type `py5jupyter` on the command line and have it install py5 kernels if needed (or ask to install the kernels if they are not installed) and then open the `jupyter console --kernel=py5` for us?

Maybe better if it checked everything and launched `jupyter lab` instead... because there you can choose a _notebook_ or a _console_ session with the `py5` kernel.

**Reply by villares** [reply]
how to check installed kernels https://stackoverflow.com/a/49299687/19771

**Reply by hx2A** [reply]
This is an interesting idea. Installing py5jupyter could create a command line utility called something like `py5console` which would install the py5 kernels (in the default locations) if they have not been installed already and then run `jupyter console --kernel=py5`. ~~One problem (I think) is that `jupyter console --kernel=py5` doesn't work at all on OSX. At least that is my recollection? I think it cannot open a sketch window.~~ Conveniently, it works just fine on OSX.

One difference between ipython and jupyter console is how it handles print statements. A running Sketch with print statements (either the builtin `print` or `py5.println`) will display output immediately with ipython but will wait for you to enter a command with jupyter console.

**Reply by villares** [reply]
Oh, does the `ipython` console use the same jupyter kernels?  

Could we have a `py5lab` sister command that opens `jupyter lab` checking if the py5 and py5bot kernels were installed?

No hurry/pressure... I'm still getting the hang of the interactive workflows.

**Comment by hx2A** [comment]
> Oh, does the `ipython` console use the same jupyter kernels?

I suspect that ipython and jupyter were developed separately, and then the jupyter console is a fork of ipython that was adapted to fit in the jupyter framework. The ipython command has no ability (that I know of) to use a different kernel besides the default Python kernel. I wouldn't be surprised if one day ipython was deprecated in favor of jupyter console.

(this is a guess, but I have spent a lot of time reading the source code for jupyter and ipython)

> Could we have a `py5lab` sister command that opens `jupyter lab` checking if the py5 and py5bot kernels were installed?

Why not? If `py5console` works, `py5lab` should work as well.

> No hurry/pressure... I'm still getting the hang of the interactive workflows.

This won't be in the next release...right now I going to work on the existing open issues and then I have a lot of documentation to write. Between the color stuff and the shapely / trimesh features, this is already a release with a lot of stuff in it!  Also, this change would have to go in py5jupyter, not py5. It is a separate Python library.

**Reply by villares** [reply]
Just an interface idea, feel free to ignore. I think the command line commands could be: `py5jupyter`. 

Without arguments it would do the kernel checking and call `jupyter lab`, where you can choose your kernel and choose between _notebook_ and _console_ (the console inside the lab interface).

![image](https://github.com/py5coding/py5generator/assets/3694604/acda9410-70ce-4836-9f68-5935c29f15ba)

![image](https://github.com/py5coding/py5generator/assets/3694604/654e92e0-af39-4ec5-bf1e-7afe9d44c46c)

And maybe `py5jupyter console`  would do the kernel checking and call `jupyter console --kernel=py5` in the terminal.

![image](https://github.com/py5coding/py5generator/assets/3694604/fd63515f-a21b-41be-a405-0f7a8ad7aeaa)

Maybe, it should also look for the JDK and ask to download it if it is missing... That way it could provide a complete streamlined installation experience.

**Reply by hx2A** [reply]
> Maybe, it should also look for the JDK and ask to download it if it is missing... That way it could provide a complete streamlined installation experience.

That's a nice touch! It would need to install install-jdk first.

Why do you get that Debugger warning about frozen modules? That shouldn't happen.

Do you use the console inside Jupyter lab? I'm not a fan of it but it does work ok.

**Reply by villares** [reply]
I have no idea of what this warning is about. I'm just ignoring it. I have used the console very little. I'm trying to get more used to these other workflows. Yesterday I was struggling with saving a notebook... My go to comfort tool is Thonny IDE.  I wish the Thonny console was more capable... using it with py5.run_sketch(block=False) can be handy for certain explorations.

**Reply by hx2A** [reply]
> My go to comfort tool is Thonny IDE. I wish the Thonny console was more capable... using it with py5.run_sketch(block=False) can be handy for certain explorations.

Can Thonny let you interact with other running Python programs in the generic Python interpreter? It is possible to run a Sketch with `python -i sketch.py` instead of `run_sketch sketch.py`. The former will execute the code in `sketch.py` and then give you the Python interpreter to enter more commands. That would have to be module mode code though but theoretically I could come up with something to make it be imported mode.

It's too bad there are not more beginner-friendly Jupyter Clients. Jupyter lab, Jupyter console, etc are all "clients" and are built to follow a "communication protocol". VSCode can also run notebooks because there is an extension to make it a Jupyter Client. There might be other clients out there that would be a good fit here. I've searched for them but haven't found anything.

**Reply by villares** [reply]
Maybe it is already the `-i` flag in action in Thonny, I'm not sure... If I was not clear, I do use the `block=False` feature to interact in the console, I know it wouldn't work on Macs with Thonny.
![image](https://github.com/py5coding/py5generator/assets/3694604/2577cc5b-d30b-4716-8095-69d8fcdb13d2)

**Reply by hx2A** [reply]
This is interactive? Interesting.

I wonder if it would be possible to do something like `run_sketch -i sketch.py`??? Can `run_sketch` support passing commands to and from a Python interpreter? This will take some thought.

**Reply by villares** [reply]
Yes, it is great fun and instructive to interact with.

**Reply by hx2A** [reply]
Interesting, I will keep that in mind as I develop ideas related to that.

**Reply by Maltiec** [reply]
that's a really mindblowing feature! ) I've tried it in Spyder's ipython console, it does not show classic REPL thing >>>, but if I run the sketch with `run -i sketch.py` it works like REPL: I can change data inside my sketch and call py5 functions.

**Reply by villares** [reply]
How cool, @Maltiec! Do post some images and more about your Spyder use. I think I installed it once with conda...

**Reply by Maltiec** [reply]
ok, I am working on game-of-life-based music gen now, I'll make some screenshots of Spyder workprocess

**Reply by hx2A** [reply]
> I can change data inside my sketch and call py5 functions.

Hi @Maltiec !

Programmatically interacting with a running Sketch is quite fun. If you like, you can even replace the `draw()` function with a new one with `hot_reload_draw()`.

https://py5coding.org/reference/sketch_hot_reload_draw.html

Be careful about calling py5 drawing methods from outside the animation thread, particularly for the OpenGL renderers. This will either be an error or cause it to crash.

**Reply by villares** [reply]
About this, being careful of what you do outside the animation thread, I was very surprised today to learn that I can't call `py5.save_frame(...)`  from my `key_pressed()` event function when using `P3D` (an OpenGL renderer)!  Was that always so? Or is is it something new to Processing 4? Also I would say it is not intuitive to think of saving an image as a "drawing method"...

```python
import py5

def setup():
    py5.size(400, 400, py5.P3D)
    py5.no_loop()
    
def draw():
    py5.circle(200, 200, 100)
    
def key_pressed():
    py5.save_frame('###.png')
    
py5.run_sketch()
``` 

Processing's warning:

```
You are trying to draw outside OpenGL's animation thread.
Place all drawing commands in the draw() function, or inside
your own functions as long as they are called from draw(),
but not in event handling functions such as keyPressed()
or mousePressed().
```

**Reply by hx2A** [reply]
> About this, being careful of what you do outside the animation thread, I was very surprised today to learn that I can't call py5.save_frame(...) from my key_pressed() event function when using P3D (an OpenGL renderer)! Was that always so? Or is is it something new to Processing 4? Also I would say it is not intuitive to think of saving an image as a "drawing method"...

Hmmm, that error message is a bit confusing.

Here's what's happening:

py5's `save_frame()` is different from Processing's `saveFrame()`. py5 uses the PIL image library to save the image, letting you save the image in many formats and all the other nice features PIL has. py5 also lets you save to a `BytesIO` object. For these work, it needs to access the frame's pixels, calling Processing's `updatePixels()` method. That call must be triggering that error.

Processing does let you call `saveFrame()` from the `mousePressed()` event with the `P3D` renderer. I just tested it because I wasn't sure what would happen.

I don't think I can get py5 to allow saving from the `mousePressed()` event but might be able to come up with something. I do know I can improve the error message to be something that isn't confusing. If you want, open an issue and I'll think about this and see what I can come up with. I guess at a minimum py5 can check to see if `save_frame()` is being called from `draw()` or not and fall back to Processing's method instead. Or at least, that's the idea that I thought of while writing this paragraph. :) Yes, so open an issue if you think this is something that is worth working on. I won't do it for this release (this weekend) but perhaps for the next release.

**Reply by villares** [reply]
Let's think about this for the future... no hurry. But intercepting this message and providing a custom py5 one would be nice.

**Reply by hx2A** [reply]
In the meantime you can use `py5_tools.screenshot()`, which would work for you here. It returns PIL Image object, which you can then save.

https://py5coding.org/reference/py5tools_screenshot.html

py5 could use this when `save_frame()` is not called from `draw()` to get the frame, then save it.

I'm going to open an issue for this because I'll forget about the issue and my ideas for fixing it.

**Reply by villares** [reply]
I'll have a look at the screenshot tool! My naive solution would be just to make a global flag variable and check it to save the frame from inside `draw()`.

**Reply by hx2A** [reply]
It happens that's exactly what the screenshot tool does. It adds a hook to your `draw()` function, attaching itself to it for one execution of `draw()` and then saves the frame.

**Comment by villares** [comment]
Another thing for people doing "interactive live coding" is somehow to simplify the sketch `hot_reload_draw` feature
https://py5coding.org/reference/sketch_hot_reload_draw.html?highlight=run_sketch 
I'm not sure how it could be done (maybe a @hot_reload decorator for when you define a new draw?), but it is something I was thinking about and I'd like to leave here so I don't forget about it.

**Reply by hx2A** [reply]
Right now the signature for the `hot_reload_draw()` method is:

```python
hot_reload_draw(
    draw: Callable,  # function to replace existing draw function
) -> None
```

I could easily make it so you don't have to pass the function. Instead it could also grab a function named `draw()` in the local namespace (which is how `run_sketch()` works...observe you don't have to explicitly pass `setup()` and `draw()` to `run_sketch()`.) Your decorator idea is interesting also. In what environment would you use this? Jupyter?

**Reply by villares** [reply]
could be Jupyter, IPython, or even the Thonny console... but I haven't played enough with them yet....
